From 47a49b91d4d439be9dee1ee62fc8f1652113bbc6 Mon Sep 17 00:00:00 2001
From: Andrei Cherechesu <andrei.cherechesu@nxp.com>
Date: Fri, 18 Mar 2022 04:46:01 +0200
Subject: [PATCH 11/11] drivers/perf: nxf-s32-ddr-perf: Check if IRQ triggered
 on counter0

Check if the IRQ handler was entered because an IRQ was triggered
by the overflow on counter0, since the same channel is shared with
the DDR ECC Scrubber interrupt.

Issue: ALB-8492
Upstream-Status: Pending 

Signed-off-by: Andrei Cherechesu <andrei.cherechesu@nxp.com>
Signed-off-by: Zhantao Tang <zhantao.tang@windriver.com>
---
 drivers/perf/nxp-s32-ddr-perf.c | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/drivers/perf/nxp-s32-ddr-perf.c b/drivers/perf/nxp-s32-ddr-perf.c
index 09eaad7c42c6..6d39ea254c90 100644
--- a/drivers/perf/nxp-s32-ddr-perf.c
+++ b/drivers/perf/nxp-s32-ddr-perf.c
@@ -434,6 +434,12 @@ static irqreturn_t ddr_perf_irq_handler(int irq, void *p)
 	struct ddr_pmu *pmu = (struct ddr_pmu *)p;
 	struct perf_event *event;
 
+	/* If no overflow on counter0, return */
+	cntl = ioread32(pmu->base + EVENT_CYCLES_COUNTER * 4 + COUNTER_CNTL);
+	cntl &= CNTL_OVER;
+	if (!cntl)
+		return IRQ_NONE;
+
 	/* all counter will stop if cycle counter disabled */
 	ddr_perf_counter_enable(pmu, EVENT_CYCLES_ID, EVENT_CYCLES_COUNTER,
 				false);
-- 
2.17.1

